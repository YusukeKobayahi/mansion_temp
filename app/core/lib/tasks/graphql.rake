# frozen_string_literal: true

module GraphqlTask
  extend Rake::DSL

  namespace :graphql do
    namespace :consumer do
      task schema_dump: :environment do
        require 'graphql/rake_task'

        GraphQL::RakeTask.new(
          load_schema: lambda { |_task|
            require File.expand_path('../../app/graphql/consumer/inno_schema', __dir__)
            Consumer::InnoSchema
          },
          directory: './graphql/consumer',
        )
        Rake::Task['graphql:schema:idl'].invoke
      end
    end
  end
end
