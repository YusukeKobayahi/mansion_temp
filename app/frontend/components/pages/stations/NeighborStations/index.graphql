fragment MsLibraryRailwayStationNeighborStations on RailwayStation {
  name
  sameStations {
    line {
      name
    }
    name
    nextStations {
      id
      name
      mansionCount
    }
  }
}
