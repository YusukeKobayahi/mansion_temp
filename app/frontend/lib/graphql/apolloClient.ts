// 参考
// https://github.com/vercel/next.js/tree/canary/examples/with-apollo

import { useMemo } from "react";
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createHttpLink } from "apollo-link-http";
import { from } from "apollo-link";
import { createFragmentArgumentLink } from "apollo-link-fragment-argument";
import { mergeDeep, relayStylePagination } from "@apollo/client/utilities";
import { isSSR } from "~/lib/utils";

/* eslint-disable @typescript-eslint/no-explicit-any -- キャッシュの型を事前に決められないため */
type CacheState = Record<string, any> | null;
/* eslint-enable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/naming-convention -- GraphQLのタイプ名 (このconventionに従わない)を指定したいので */
const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Mansion: {
        keyFields: ["uniqueCode"],
        fields: {
          salesHistories: relayStylePagination(),
        },
      },
    },
  });
/* eslint-enable @typescript-eslint/naming-convention */

const link = from([
  createFragmentArgumentLink(),
  createHttpLink({
    uri: `${process.env.NEXT_PUBLIC_CORE_API_URL}/consumer/graphql`,
    fetchOptions: {
      mode: "cors",
    },
  }),
]);

function createApolloClient() {
  return new ApolloClient({
    ssrMode: isSSR(),
    // See: https://github.com/apollographql/apollo-link/issues/1258#issuecomment-604031488
    // TODO: @apollo/clientバージョン上げたら直るかも
    link: (link as unknown) as ApolloLink,
    cache: createCache(),
    defaultOptions: {
      watchQuery: {
        // SSR時にuseQueryによってリクエストしてレンダリングする場合、
        // getServerSidePropsからキャッシュが渡っている必要がある。(でないと描画されない)
        // そのため、SSR時は常にcacheから取得されるべきで、実装意図を明確にするためPolicyも設定しておく
        // (エラーになるわけではない)
        fetchPolicy: isSSR() ? "cache-only" : "cache-first",
      },
    },
  });
}

let apolloClient: ReturnType<typeof createApolloClient>;

export function initializeApollo(
  initialState: CacheState = null
): ApolloClient<CacheState> {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the existing cache into data passed from getStaticProps/getServerSideProps
    const data = mergeDeep(initialState, existingCache);

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (isSSR()) return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState: CacheState): ApolloClient<CacheState> {
  const store = useMemo(() => initializeApollo(initialState), [initialState]);
  return store;
}
