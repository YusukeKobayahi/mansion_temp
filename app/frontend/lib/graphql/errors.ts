export interface GraphQLError {
  graphQLErrors?: {
    message: string;
    extensions?: {
      code: string;
    };
  }[];
}

export const isNotFoundError = (err: GraphQLError): boolean => {
  const graphQLErrors = err.graphQLErrors;
  if (graphQLErrors == null) return false;

  return graphQLErrors[0]?.extensions?.code === "NOT_FOUND";
};
