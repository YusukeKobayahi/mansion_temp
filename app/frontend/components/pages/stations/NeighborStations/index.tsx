import { MsLibraryRailwayStationNeighborStationsFragment } from "~/graphql/generated";

import styles from "~/components/pages/towns/IndexedTowns/index.module.scss";
import BoxHeading from "~/components/commons/headings/BoxHeading";
import LinkBlock from "~/components/commons/LinkBlock";
import flatten from "lodash/flatten";
import uniqBy from "lodash/uniqBy";

export interface NeighborStationsProps {
  data: MsLibraryRailwayStationNeighborStationsFragment;
}

const NeighborStations: React.FC<NeighborStationsProps> = ({
  data,
}: NeighborStationsProps) => {
  if (
    !data.sameStations ||
    data.sameStations.length === 0 ||
    data.sameStations.every((station) => station.nextStations.length === 0)
  )
    return null;

  const uniqNextStations = uniqBy(
    flatten(
      data.sameStations.map((sameStations) => {
        return sameStations.nextStations.map((station) => station);
      })
    ),
    (item) => item.name
  );

  const items = uniqNextStations.map((station) => {
    return (
      <LinkBlock
        key={station.id}
        name={station.name + "駅"}
        id={station.id}
        link="stations"
        mansionCount={station.mansionCount}
        eventParams={{
          action: "msl_area_click",
          category: "msl_ichiran",
          label: "msl_linked_cities",
          value: 1,
        }}
      />
    );
  });

  return (
    <section className={styles.container}>
      <h2>
        <BoxHeading>
          {data.name + "駅"}
          の乗り入れ路線の隣接駅から中古マンションを探す
        </BoxHeading>
      </h2>
      <ul className={styles.list}>{items}</ul>
    </section>
  );
};

export default NeighborStations;
