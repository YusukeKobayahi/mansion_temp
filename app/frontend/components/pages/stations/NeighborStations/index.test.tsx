import { render } from "@testing-library/react";
import mockData from "~/components/pages/stations/fixtures/mockData.json";
import { StationsQuery } from "~/graphql/generated";
import NeighborStations from "~/components/pages/stations/NeighborStations";

describe("components/pages/stations/NeighborStations", () => {
  it("正常に表示される", () => {
    const mock = mockData as StationsQuery;
    const { asFragment } = render(<NeighborStations data={mock.stations[0]} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
