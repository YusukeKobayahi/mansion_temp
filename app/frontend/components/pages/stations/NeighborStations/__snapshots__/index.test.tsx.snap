// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/stations/NeighborStations 正常に表示される 1`] = `
<DocumentFragment>
  <section
    class="container"
  >
    <h2>
      <div
        class="heading"
      >
        横浜駅の乗り入れ路線の隣接駅から中古マンションを探す
      </div>
    </h2>
    <ul
      class="list"
    >
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/1321"
        >
          新橋駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/1323"
        >
          川崎駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/1368"
        >
          田町駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/1370"
        >
          高輪ゲートウェイ駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/1475"
        >
          西大井駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/1934"
        >
          渋谷駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/1936"
        >
          東京駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/2002"
        >
          大井町駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/5727"
        >
          泉岳寺駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
      <li
        class="item"
      >
        <button
          class="link"
          href="/stations/5729"
        >
          北品川駅
          <span
            class="count"
          >
            (undefined件)
          </span>
        </button>
      </li>
    </ul>
  </section>
</DocumentFragment>
`;
